File: app/src/main/java/org/bepass/oblivion/utils/ThemeHelper.java
package org.bepass.oblivion.utils;

import android.view.View;

import androidx.appcompat.app.AppCompatDelegate;

public class ThemeHelper {

    public void updateActivityBackground(View root) {
    }

    public enum Theme {
        LIGHT(AppCompatDelegate.MODE_NIGHT_NO, AppCompatDelegate.MODE_NIGHT_NO),
        DARK(AppCompatDelegate.MODE_NIGHT_YES, AppCompatDelegate.MODE_NIGHT_YES),
        FOLLOW_SYSTEM(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM, AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM),
        UNSPECIFIED(AppCompatDelegate.MODE_NIGHT_UNSPECIFIED, AppCompatDelegate.MODE_NIGHT_UNSPECIFIED),
        OLED(3, AppCompatDelegate.MODE_NIGHT_YES);

        private final int storageValue;
        private final int nightMode;

        Theme(int storageValue, int nightMode) {
            this.storageValue = storageValue;
            this.nightMode = nightMode;
        }

        public int getStorageValue() {
            return storageValue;
        }

        @AppCompatDelegate.NightMode
        public int getNightMode() {
            return nightMode;
        }

        public static Theme fromStorageKey(int storageValue) {
            for (Theme theme : values()) {
                if (theme.getStorageValue() == storageValue) {
                    return theme;
                }
            }
            return FOLLOW_SYSTEM;
        }
    }

    private static volatile ThemeHelper instance;
    private Theme currentTheme = Theme.FOLLOW_SYSTEM;

    private ThemeHelper() {
        // Private constructor for Singleton
    }

    public static ThemeHelper getInstance() {
        if (instance == null) {
            synchronized (ThemeHelper.class) {
                if (instance == null) {
                    instance = new ThemeHelper();
                }
            }
        }
        return instance;
    }

    public void init() {
        int themeMode = FileManager.getInt(FileManager.KeyHolder.DARK_MODE);
        // If key doesn't exist or returns invalid, default to FOLLOW_SYSTEM
        if (themeMode == -1 && !FileManager.contains(FileManager.KeyHolder.DARK_MODE)) {
            // Check if it was never set, default to FOLLOW_SYSTEM
            currentTheme = Theme.FOLLOW_SYSTEM;
        } else {
            currentTheme = Theme.fromStorageKey(themeMode);
        }
        applyTheme();
    }

    public void applyTheme() {
        AppCompatDelegate.setDefaultNightMode(currentTheme.getNightMode());
    }

    public void select(Theme theme) {
        currentTheme = theme;
        FileManager.set(FileManager.KeyHolder.DARK_MODE, theme.storageValue);
        applyTheme();
    }

    public Theme getCurrentTheme() {
        return currentTheme;
    }
    
    public boolean isOled() {
        return currentTheme == Theme.OLED;
    }

}

File: app/src/main/java/org/bepass/oblivion/utils/ThemeHelper.kt
package org.bepass.oblivion.utils

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

object ThemeHelper {
  private const val LIGHT_STORAGE_VALUE = 1
  private const val DARK_STORAGE_VALUE = 2
  private const val OLED_STORAGE_VALUE = 3
  const val MODE_NIGHT_FOLLOW_SYSTEM = -1
  const val MODE_NIGHT_UNSPECIFIED = -100
  const val MODE_NIGHT_NO = 1
  const val MODE_NIGHT_YES = 2

  enum class Theme(val storageValue: Int, val nightMode: Int) {
    OLED(OLED_STORAGE_VALUE, MODE_NIGHT_YES),
    DARK(DARK_STORAGE_VALUE, MODE_NIGHT_YES),
    LIGHT(LIGHT_STORAGE_VALUE, MODE_NIGHT_NO);

    companion object {
      fun fromStorageKey(storageValue: Int): Theme =
        entries.firstOrNull { it.storageValue == storageValue } ?: OLED
    }
  }

  @JvmStatic fun getInstance(): ThemeHelper = this

  private val _themeFlow = MutableStateFlow(Theme.OLED)
  val themeFlow: StateFlow<Theme> = _themeFlow.asStateFlow()

  var currentTheme: Theme = Theme.OLED
    private set

  fun init() {
    val storedValue = FileManager.getInt(FileManager.Keys.DARK_MODE, Theme.OLED.storageValue)
    currentTheme = Theme.fromStorageKey(storedValue)
    _themeFlow.value = currentTheme
    if (!FileManager.contains(FileManager.Keys.DARK_MODE) || storedValue != currentTheme.storageValue) {
      FileManager.set(FileManager.Keys.DARK_MODE, currentTheme.storageValue)
    }
    applyTheme()
  }

  fun applyTheme() = Unit

  fun select(theme: Theme) {
    currentTheme = theme
    _themeFlow.value = theme
    FileManager.set(FileManager.Keys.DARK_MODE, theme.storageValue)
    applyTheme()
  }

  val isOled: Boolean
    get() = currentTheme == Theme.OLED

  fun isDarkTheme(@Suppress("UNUSED_PARAMETER") systemDark: Boolean): Boolean = currentTheme != Theme.LIGHT
}
